File: packages/dvai-bridge-ios/ios/Sources/DVAIBridge/Capability/DeviceID.swift
import Foundation

/// Stable per-install device identifier. Generated once on first call,
/// persisted alongside the capability cache under the same Application
/// Support directory. Used for:
///   - identifying THIS device in mDNS TXT records (LAN discovery).
///   - identifying THIS device in rendezvous-server pairing payloads.
///   - keying the capability cache.
///
/// NOT a privacy hazard: the ID is per-install and per-device-storage,
/// never tied to user identity. Reinstalling the app or wiping app
/// storage produces a fresh ID — that's the right behaviour and matches
/// the TS-side `generateDeviceId` semantics in
/// `packages/dvai-bridge-core/src/capability/deviceId.ts`.
public final class DeviceIDStore: @unchecked Sendable {
    private let fileURL: URL
    private let lock = NSLock()
    private var cached: String?

    public init(directory: URL) {
        self.fileURL = directory.appendingPathComponent("device-id.txt", isDirectory: false)
    }

    /// Return the device ID, generating + persisting it on first call.
    public func get() throws -> String {
        lock.lock()
        defer { lock.unlock() }
        if let cached = cached { return cached }

        // Try to load existing
        if let data = try? Data(contentsOf: fileURL),
           let raw = String(data: data, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines),
           !raw.isEmpty {
            cached = raw
            return raw
        }

        // Generate fresh
        let id = Self.generate()
        try FileManager.default.createDirectory(
            at: fileURL.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )
        try id.write(to: fileURL, atomically: true, encoding: .utf8)
        cached = id
        return id
    }

    /// Generate a 22-char URL-safe base64 random ID. Mirrors the TS
    /// `generateDeviceId()` shape (16 random bytes, base64url, no
    /// padding) so device IDs round-trip across platforms.
    public static func generate() -> String {
        var bytes = [UInt8](repeating: 0, count: 16)
        let status = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        if status != errSecSuccess {
            // Fall back to UUID — still random, just not from SecRandom.
            return UUID().uuidString.replacingOccurrences(of: "-", with: "")
                .lowercased().prefix(22).description
        }
        return base64UrlEncode(Data(bytes))
    }

    private static func base64UrlEncode(_ data: Data) -> String {
        let b64 = data.base64EncodedString()
        return b64
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}

File: packages/dvai-bridge-android-shared-core/android/src/main/java/co/deepvoiceai/bridge/shared/core/capability/DeviceID.kt
package co.deepvoiceai.bridge.shared.core.capability

import android.content.Context
import java.io.File
import java.security.SecureRandom
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Stable per-install device identifier (Kotlin mirror of
 * `@dvai-bridge/core/src/capability/deviceId.ts`).
 *
 * - Generated once on first call, then persisted under
 *   `applicationContext.cacheDir/dvai-bridge/device.json`.
 * - Used to identify THIS device in mDNS TXT records, rendezvous
 *   pairing payloads, and as the key for the capability cache.
 *
 * NOT a privacy hazard: the ID is per-install and per-cache-storage,
 * never tied to user identity. Reinstalling the app or wiping app
 * storage produces a fresh ID — that's the right behaviour.
 */
object DeviceID {
    private const val FILENAME = "device.json"
    private const val CACHE_DIR_NAME = "dvai-bridge"

    @Serializable
    private data class DeviceFile(val deviceId: String)

    private val json = Json { ignoreUnknownKeys = true; prettyPrint = true }

    /**
     * Read the persistent device ID, generating + persisting one if
     * none exists. Safe to call from any thread; the file write
     * is atomic-enough for the per-install lifetime guarantee
     * (single-writer, never concurrent across processes).
     */
    @Synchronized
    fun get(context: Context): String {
        val dir = File(context.applicationContext.cacheDir, CACHE_DIR_NAME)
        if (!dir.exists()) dir.mkdirs()
        val file = File(dir, FILENAME)
        if (file.exists()) {
            try {
                val txt = file.readText(Charsets.UTF_8)
                val parsed = json.decodeFromString(DeviceFile.serializer(), txt)
                if (parsed.deviceId.isNotEmpty()) return parsed.deviceId
            } catch (_: Throwable) {
                // Fall through and regenerate on parse error.
            }
        }
        val fresh = generate()
        try {
            file.writeText(json.encodeToString(DeviceFile.serializer(), DeviceFile(fresh)), Charsets.UTF_8)
        } catch (_: Throwable) {
            // If we can't persist, the caller still gets a valid ID for
            // this process; the next call will try again.
        }
        return fresh
    }

    /**
     * Reset the persisted device ID. Returns the new ID. Mainly for
     * tests + the "forget all paired peers" UI.
     */
    @Synchronized
    fun reset(context: Context): String {
        val dir = File(context.applicationContext.cacheDir, CACHE_DIR_NAME)
        if (!dir.exists()) dir.mkdirs()
        val file = File(dir, FILENAME)
        val fresh = generate()
        file.writeText(json.encodeToString(DeviceFile.serializer(), DeviceFile(fresh)), Charsets.UTF_8)
        return fresh
    }

    /** Generate a 22-char URL-safe random ID. Equivalent to the TS impl. */
    fun generate(): String {
        val bytes = ByteArray(16)
        SecureRandom().nextBytes(bytes)
        return base64UrlEncode(bytes)
    }

    private fun base64UrlEncode(bytes: ByteArray): String {
        val sb = StringBuilder()
        val b64 = java.util.Base64.getEncoder().encodeToString(bytes)
        for (c in b64) {
            when (c) {
                '+' -> sb.append('-')
                '/' -> sb.append('_')
                '=' -> Unit // strip padding
                else -> sb.append(c)
            }
        }
        return sb.toString()
    }
}
